File: backend/app/schemas/orchestration.py
"""Pydantic request and response schemas for Phase 34 API Layer."""

from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict


# 1. Workflows
class StartWorkflowRequest(BaseModel):
    workflow_key: str
    input_data: Optional[Dict[str, Any]] = None
    trigger_type: str = "MANUAL"
    trigger_reference: Optional[str] = None


class WorkflowRunStepResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    step_key: str
    step_type: str
    status: str
    attempt_count: int
    duration_ms: float
    input_data: Dict[str, Any]
    output_data: Dict[str, Any]
    error_code: Optional[str] = None
    error_summary: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None


class WorkflowRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    workflow_key: str
    workflow_version: str
    trigger_type: str
    trigger_reference: Optional[str] = None
    status: str
    current_step: Optional[str] = None
    correlation_id: str
    input_data: Dict[str, Any]
    output_data: Dict[str, Any]
    error_code: Optional[str] = None
    error_summary: Optional[str] = None
    started_at: datetime
    completed_at: Optional[datetime] = None
    paused_at: Optional[datetime] = None
    failed_at: Optional[datetime] = None
    cancelled_at: Optional[datetime] = None


class WorkflowTemplateResponse(BaseModel):
    workflow_key: str
    name: str
    description: str
    category: str
    version: str
    steps: List[Dict[str, Any]]
    transitions: List[Dict[str, Any]]


# 2. Events
class EventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    event_id: str
    event_type: str
    event_version: str
    aggregate_type: str
    aggregate_id: str
    tenant_id: str
    payload: Dict[str, Any]
    correlation_id: str
    causation_id: Optional[str] = None
    status: str
    attempt_count: int
    created_at: datetime


class EventReplayRequest(BaseModel):
    event_type: str
    replay_mode: str = "DRY_RUN"  # READ_ONLY, DRY_RUN, REBUILD_PROJECTION, CONTROLLED_REEXECUTION


class EventReplayResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    event_type: str
    replay_mode: str
    events_replayed_count: int
    triggered_by: str
    status: str
    created_at: datetime


# 3. Human Tasks
class CompleteHumanTaskRequest(BaseModel):
    decision: str  # APPROVED, REJECTED, REVISION_REQUIRED
    decision_reason: str
    content_hash: Optional[str] = None


class HumanTaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    workflow_run_id: str
    step_key: str
    title: str
    description: str
    task_type: str
    priority: str
    assigned_to: Optional[str] = None
    status: str
    deadline: Optional[datetime] = None
    input_data: Dict[str, Any]
    decision: Optional[str] = None
    decision_reason: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None


# 4. Automations
class CreateAutomationRuleRequest(BaseModel):
    name: str
    description: str
    trigger_type: str = "EVENT"
    trigger_config: Dict[str, Any]
    condition_config: Dict[str, Any]
    action_config: Dict[str, Any]
    requires_human_approval: bool = True


class AutomationRuleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    name: str
    description: str
    version: str
    trigger_type: str
    trigger_config: Dict[str, Any]
    condition_config: Dict[str, Any]
    action_config: Dict[str, Any]
    priority: str
    enabled: bool
    requires_human_approval: bool
    created_by: str
    created_at: datetime
    updated_at: datetime


# 5. Dead Letter Queue
class DeadLetterResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    event_id: str
    event_type: str
    workflow_id: Optional[str] = None
    consumer: str
    attempt_count: int
    failure_type: str
    error_summary: str
    payload: Dict[str, Any]
    status: str
    last_error_at: datetime
    created_at: datetime

File: frontend/lib/api/orchestration.ts
import { api } from './client';

export interface WorkflowTemplate {
  workflow_key: string;
  name: string;
  description: string;
  category: string;
  version: string;
  steps: Array<{
    step_key: string;
    name: string;
    step_type: string;
    config: Record<string, any>;
    timeout_seconds: number;
  }>;
  transitions: Array<{
    from_step_key: string;
    to_step_key: string;
    condition?: string;
    is_default: boolean;
  }>;
}

export interface WorkflowRunStep {
  id: string;
  step_key: string;
  step_type: string;
  status: string;
  attempt_count: number;
  duration_ms: number;
  input_data: Record<string, any>;
  output_data: Record<string, any>;
  error_code?: string;
  error_summary?: string;
  started_at?: string;
  completed_at?: string;
}

export interface WorkflowRun {
  id: string;
  tenant_id: string;
  workflow_key: string;
  workflow_version: string;
  trigger_type: string;
  trigger_reference?: string;
  status: string;
  current_step?: string;
  correlation_id: string;
  input_data: Record<string, any>;
  output_data: Record<string, any>;
  error_code?: string;
  error_summary?: string;
  started_at: string;
  completed_at?: string;
  paused_at?: string;
  failed_at?: string;
  cancelled_at?: string;
}

export interface DomainEventRecord {
  id: string;
  event_id: string;
  event_type: string;
  event_version: string;
  aggregate_type: string;
  aggregate_id: string;
  tenant_id: string;
  payload: Record<string, any>;
  correlation_id: string;
  causation_id?: string;
  status: string;
  attempt_count: number;
  created_at: string;
}

export interface HumanTask {
  id: string;
  tenant_id: string;
  workflow_run_id: string;
  step_key: string;
  title: string;
  description: string;
  task_type: string;
  priority: string;
  assigned_to?: string;
  status: string;
  deadline?: string;
  input_data: Record<string, any>;
  decision?: string;
  decision_reason?: string;
  created_at: string;
  completed_at?: string;
}

export interface AutomationRule {
  id: string;
  tenant_id: string;
  name: string;
  description: string;
  version: string;
  trigger_type: string;
  trigger_config: Record<string, any>;
  condition_config: Record<string, any>;
  action_config: Record<string, any>;
  priority: string;
  enabled: boolean;
  requires_human_approval: boolean;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface DeadLetterRecord {
  id: string;
  event_id: string;
  event_type: string;
  workflow_id?: string;
  consumer: string;
  attempt_count: number;
  failure_type: string;
  error_summary: string;
  payload: Record<string, any>;
  status: string;
  last_error_at: string;
  created_at: string;
}

export const orchestrationApi = {
  // Workflows
  getTemplates: async (): Promise<WorkflowTemplate[]> => {
    return api<WorkflowTemplate[]>('/workflows/templates');
  },
  listWorkflows: async (status?: string, workflow_key?: string): Promise<WorkflowRun[]> => {
    const params = new URLSearchParams();
    if (status) params.append('status', status);
    if (workflow_key) params.append('workflow_key', workflow_key);
    const query = params.toString() ? `?${params.toString()}` : '';
    return api<WorkflowRun[]>(`/workflows${query}`);
  },
  getWorkflow: async (runId: string): Promise<WorkflowRun> => {
    return api<WorkflowRun>(`/workflows/${runId}`);
  },
  startWorkflow: async (data: {
    workflow_key: string;
    input_data?: Record<string, any>;
    trigger_type?: string;
    trigger_reference?: string;
  }): Promise<WorkflowRun> => {
    return api<WorkflowRun>('/workflows/start', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },
  pauseWorkflow: async (runId: string): Promise<WorkflowRun> => {
    return api<WorkflowRun>(`/workflows/${runId}/pause`, {
      method: 'POST',
    });
  },
  resumeWorkflow: async (runId: string): Promise<WorkflowRun> => {
    return api<WorkflowRun>(`/workflows/${runId}/resume`, {
      method: 'POST',
    });
  },
  cancelWorkflow: async (runId: string, reason?: string): Promise<WorkflowRun> => {
    return api<WorkflowRun>(`/workflows/${runId}/cancel?reason=${encodeURIComponent(reason || 'Cancelled by user')}`, {
      method: 'POST',
    });
  },

  // Events
  listEvents: async (event_type?: string): Promise<DomainEventRecord[]> => {
    const params = new URLSearchParams();
    if (event_type) params.append('event_type', event_type);
    const query = params.toString() ? `?${params.toString()}` : '';
    return api<DomainEventRecord[]>(`/events${query}`);
  },
  replayEvents: async (data: { event_type: string; replay_mode?: string }): Promise<any> => {
    return api<any>('/events/replay', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  // Human Tasks
  listTasks: async (status?: string): Promise<HumanTask[]> => {
    const params = new URLSearchParams();
    if (status) params.append('status', status);
    const query = params.toString() ? `?${params.toString()}` : '';
    return api<HumanTask[]>(`/tasks${query}`);
  },
  completeTask: async (taskId: string, data: { decision: string; decision_reason: string; content_hash?: string }): Promise<HumanTask> => {
    return api<HumanTask>(`/tasks/${taskId}/complete`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },

  // Automations
  listAutomations: async (enabled_only?: boolean): Promise<AutomationRule[]> => {
    const params = new URLSearchParams();
    if (enabled_only) params.append('enabled_only', 'true');
    const query = params.toString() ? `?${params.toString()}` : '';
    return api<AutomationRule[]>(`/automations${query}`);
  },
  createAutomation: async (data: Partial<AutomationRule>): Promise<AutomationRule> => {
    return api<AutomationRule>('/automations', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  },
  toggleAutomation: async (ruleId: string, enabled: boolean): Promise<AutomationRule> => {
    return api<AutomationRule>(`/automations/${ruleId}/toggle?enabled=${enabled}`, {
      method: 'POST',
    });
  },

  // DLQ
  listDeadLetters: async (status?: string): Promise<DeadLetterRecord[]> => {
    const params = new URLSearchParams();
    if (status) params.append('status', status);
    const query = params.toString() ? `?${params.toString()}` : '';
    return api<DeadLetterRecord[]>(`/dead-letter${query}`);
  },
  retryDeadLetter: async (dlqId: string): Promise<DeadLetterRecord> => {
    return api<DeadLetterRecord>(`/dead-letter/${dlqId}/retry`, {
      method: 'POST',
    });
  },
};
